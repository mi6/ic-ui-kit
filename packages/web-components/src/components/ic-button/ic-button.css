@import "../../global/normalize.css";

/**
* @prop --min-width: Minimum width of the button.
*/

:host {
  display: inline-block;

  --button-default: var(--ic-action-default);
  --button-default-hover: var(--ic-action-default-hover);
  --button-default-active: var(--ic-action-default-active);
  --button-default-background-hover: var(--ic-action-default-bg-hover);
  --button-default-background-active: var(--ic-action-default-bg-active);
  --icon-width: 100%;
  --icon-height: 100%;
}

.button {
  font-family: var(--ic-font-body-family);
  text-decoration: none;
  font-weight: 600;
  font-size: 0.875rem;
  transition: var(--ic-easing-transition-fast);
  border-radius: var(--ic-border-radius);
  min-width: var(--min-width, 6.25rem);
  display: inline-flex;
  gap: var(--ic-space-xs);
  flex-direction: row;
  justify-content: center;
  align-items: center;
  background: none;
  border: none;
  box-sizing: border-box;
  white-space: nowrap;
  vertical-align: middle;
}

.button:hover {
  cursor: pointer;
}

.button:focus,
:host .button:focus {
  box-shadow: var(--ic-border-focus);
}

:host(.light) .button:focus {
  box-shadow: var(--ic-border-focus);
}

.button:focus-visible {
  outline: var(--ic-hc-focus-outline);
}

:host(.dark) .button {
  --button-default: var(--ic-action-dark);
  --button-default-hover: var(--ic-action-dark);
  --button-default-active: var(--ic-action-dark);
  --button-default-background-hover: var(--ic-action-dark-bg-hover);
  --button-default-background-active: var(--ic-action-dark-bg-active);
}

:host(.light) .button {
  --button-default: var(--ic-action-light);
  --button-default-hover: var(--ic-action-light);
  --button-default-active: var(--ic-action-light);
  --button-default-background-hover: var(--ic-action-light-bg-hover);
  --button-default-background-active: var(--ic-action-light-bg-active);
}

:host(.button-variant-primary.light) .button {
  color: var(--ic-color-primary-text);

  --button-default-hover: var(--ic-action-light-hover);
  --button-default-active: var(--ic-action-light-active);
}

:host(.button-variant-primary.dark) .button {
  --button-default-hover: var(--ic-action-dark-hover);
  --button-default-active: var(--ic-action-dark-active);
}

:host(.disabled),
:host(.disabled) .button,
:host(.loading),
:host(.loading) .button {
  pointer-events: none;
}

:host(.loading-with-icon) .button {
  min-width: var(--min-width, 8.25rem);
}

/* Variants */

/* Primary */

:host(.button-variant-primary) .button {
  color: var(--ic-architectural-white);
  background-color: var(--button-default);
}

:host(.button-variant-primary) .button:hover:not(:focus) {
  background-color: var(--button-default-hover);
}

:host(.button-variant-primary) .button:active:not(:focus),
:host(.button-variant-primary.loading) .button {
  background-color: var(--button-default-active);
}

:host(.button-variant-primary.disabled) .button {
  background: var(--ic-architectural-200);
  color: var(--ic-architectural-300);
}

:host(.button-variant-primary.light.disabled) .button {
  background: var(--ic-architectural-600);
  color: var(--ic-architectural-500);
}

/* Secondary */

:host(.button-variant-secondary) .button {
  border: var(--ic-space-1px) solid var(--button-default);
  color: var(--button-default);
}

:host(.button-variant-secondary) .button:hover:not(:focus) {
  background-color: var(--button-default-background-hover);
  border-color: var(--button-default-hover);
  color: var(--button-default-hover);
}

:host(.button-variant-secondary) .button:active:not(:focus) {
  border-color: var(--button-default-active);
  background-color: var(--button-default-background-active);
  color: var(--button-default-active);
}

:host(.button-variant-secondary.loading) .button {
  border-color: var(--button-default-active);
  background-color: var(--button-default-background-active);
  color: var(--button-default-active);
}

:host(.button-variant-secondary.disabled) .button,
:host(.button-variant-secondary.disabled) .button:hover,
:host(.button-variant-secondary.disabled) .button:active {
  border-color: var(--ic-architectural-300);
  color: var(--ic-architectural-300);
  background: none;
}

:host(.button-variant-secondary.light.disabled) .button,
:host(.button-variant-secondary.light.disabled) .button:hover,
:host(.button-variant-secondary.light.disabled) .button:active {
  border-color: var(--ic-architectural-500);
  color: var(--ic-architectural-500);
  background: none;
}

/* Tertiary */

:host(.button-variant-tertiary) .button {
  color: var(--button-default);
}

:host(.button-variant-tertiary) .button:hover:not(:focus) {
  background-color: var(--button-default-background-hover);
  color: var(--button-default-hover);
}

:host(.button-variant-tertiary) .button:active:not(:focus),
:host(.button-variant-tertiary.loading) .button {
  background-color: var(--button-default-background-active);
  color: var(--button-default-active);
}

:host(.button-variant-tertiary.disabled) .button,
:host(.button-variant-tertiary.disabled) .button:hover,
:host(.button-variant-tertiary.disabled) .button:active {
  border-color: var(--ic-architectural-300);
  color: var(--ic-architectural-300);
  background: none;
}

:host(.button-variant-tertiary.light.disabled) .button,
:host(.button-variant-tertiary.light.disabled) .button:hover,
:host(.button-variant-tertiary.light.disabled) .button:active {
  border-color: var(--ic-architectural-500);
  color: var(--ic-architectural-500);
  background: none;
}

/* Icon */

:host(.button-variant-icon) .button {
  color: var(--button-default);
  min-width: 0;
}

:host(.button-variant-icon) .button:hover:not(:focus) {
  background-color: var(--button-default-background-hover);
  color: var(--button-default-hover);
}

:host(.button-variant-icon) .button:active:not(:focus),
:host(.button-variant-icon.loading) .button {
  background-color: var(--button-default-background-active);
  color: var(--button-default-active);
}

:host(.button-variant-icon.disabled) .button,
:host(.button-variant-icon.disabled) .button:hover,
:host(.button-variant-icon.disabled) .button:active {
  color: var(--ic-architectural-300);
  background: none;
}

/* Destructive */

:host(.button-variant-destructive) .button {
  color: var(--ic-architectural-white);
  background-color: var(--ic-action-destructive);
  text-transform: uppercase;
}

:host(.button-variant-destructive) .button:hover:not(:focus) {
  background-color: var(--ic-action-destructive-hover);
}

:host(.button-variant-destructive) .button:active:not(:focus),
:host(.button-variant-destructive.loading) .button {
  background-color: var(--ic-action-destructive-active);
}

:host(.button-variant-destructive.disabled) .button {
  background: var(--ic-architectural-200);
  color: var(--ic-architectural-300);
}

/* Sizing */

:host(.button-size-default) .button {
  height: 2.5rem;
  padding: var(--ic-space-xs) var(--ic-space-md);
}

:host(.button-size-small) .button {
  height: var(--ic-space-xl);
  padding: var(--ic-space-xxs) var(--ic-space-md);
}

:host(.button-size-large) .button {
  height: var(--ic-space-xxl);
  padding: var(--ic-space-sm) var(--ic-space-md);
}

:host(.button-size-default.button-variant-icon) .button {
  height: var(--ic-space-xl);
  width: var(--ic-space-xl);
  padding: 0.375rem;
}

:host(.button-size-small.button-variant-icon) .button {
  height: var(--ic-space-lg);
  width: var(--ic-space-lg);
  padding: var(--ic-space-xxs);
}

:host(.button-size-large.button-variant-icon) .button {
  height: 2.5rem;
  width: 2.5rem;
  padding: var(--ic-space-xs);
}

/* Width */
:host(.full-width),
:host(.full-width) .button {
  width: 100%;
}

/* Loading */

div.loading-container {
  position: relative;
  align-items: center;
  width: 100%;
}

ic-loading-indicator {
  --outer-color: transparent;
}

@keyframes loading-animation {
  0% {
    width: 0%;
    left: 0%;
  }

  25% {
    width: 0%;
    left: 0%;
    opacity: 0;
  }

  50% {
    width: 100%;
    left: 0%;
    opacity: 1;
  }

  75% {
    width: 0%;
    left: 100%;
  }

  100% {
    width: 0%;
    left: 100%;
    opacity: 0;
  }
}

/* Icons */

div.icon-container {
  box-sizing: border-box;
  width: var(--ic-space-lg);
  height: var(--ic-space-lg);
  display: flex;
  justify-content: center;
  align-items: center;
}

::slotted(*) {
  width: var(--icon-width) !important;
  height: var(--icon-height) !important;
  fill: currentcolor !important;
  pointer-events: none;
}

:host(.button-variant-icon) .button .icon-container {
  margin: 0;
  pointer-events: none;
}

/** SEARCH **/

:host(.search-submit-button) ::slotted(svg) {
  --icon-height: 1.25rem;
  --icon-width: 1.25rem;
}

:host(.search-submit-button-small) ::slotted(svg) {
  --icon-height: 1rem;
  --icon-width: 1rem;
}

:host(.clear-button) {
  margin: 0 var(--ic-space-xxs);
}

:host(.clear-button) .button:focus {
  box-shadow: none;
}

:host(.search-submit-button) .button:focus {
  box-shadow: none;
}

:host(.search-submit-button) .button:not(:active):focus {
  box-shadow: none;
  background-color: var(--ic-action-default-bg-hover);
}

:host(.search-submit-button) {
  display: flex;
  align-items: center;
  margin: 0 var(--ic-space-xxs);
}

:host(.menu-close-button) ::slotted(svg) {
  --icon-height: 0.875rem;
  --icon-width: 0.875rem;
}

:host(.popout-menu-button) .button {
  justify-content: left;
  border-radius: 0;
  white-space: pre-line;
  text-align: start;
}

:host(.popout-menu-button) div.icon-container {
  flex: none;
}

:host(.popout-menu-button) .button:focus {
  box-shadow: var(--ic-border-focus-inset);
  border-radius: var(--ic-border-radius-inset);
}

.ic-button-describedby {
  display: none;
}

:host .ic-tooltip {
  display: block;
}

/** High Contrast **/
@media (forced-colors: active) {
  .button {
    border: 0.125rem solid transparent;
  }
}

/** Pagination **/

:host(.flip) ::slotted(svg) {
  transform: scaleX(-1);
}
