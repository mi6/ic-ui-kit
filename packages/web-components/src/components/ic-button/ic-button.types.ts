export type IcButtonTypes = "reset" | "submit" | "button";

export type IcButtonVariants =
  | "primary"
  | "secondary"
  | "tertiary"
  | "icon"
  | "destructive";

export type IcButtonTooltipPlacement = "top" | "right" | "bottom" | "left";
