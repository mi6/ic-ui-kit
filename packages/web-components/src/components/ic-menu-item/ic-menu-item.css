@import "../../global/normalize.css";

li {
  list-style: none;
  padding: 0;
}

:host::part(button) {
  color: var(--ic-color-primary-text);
  padding: calc(var(--ic-space-xxxs) / 2);
  display: flex;
  align-items: flex-start;
  gap: var(--ic-space-xs);
  border-radius: 0;
  min-height: 2.5rem;
  height: auto;
}

.focus-border {
  display: flex;
  align-items: flex-start;
  gap: var(--ic-space-xs);
  border-radius: var(--ic-border-radius);
  margin: calc(var(--ic-space-xxxs) / 2);
  padding: var(--ic-space-xxs);
  border: var(--ic-space-xxxs) dashed transparent;

  --icon-height: auto;
}

:host(.disabled)::part(button) {
  color: var(--ic-architectural-200) !important;
  pointer-events: none;
}

:host(.disabled) .menu-item-label,
:host(.disabled) .shortcut,
:host(.disabled) .menu-item-description {
  color: var(--ic-architectural-200) !important;
}

:host([variant="destructive"])::part(button) {
  color: var(--ic-action-destructive);
}

:host::part(button):hover {
  background-color: var(--ic-action-dark-bg-hover);
}

:host::part(button):focus-visible,
:host::part(button):focus {
  color: var(--ic-color-white-text);
  background-color: var(--ic-focus-blue);
  box-shadow: none;
}

:host(.disabled)::part(button):focus-visible,
:host(.disabled)::part(button):focus {
  color: var(--ic-theme-lighten-40) !important;
}

:host([variant="destructive"])::part(button):focus-visible,
:host([variant="destructive"])::part(button):focus {
  background-color: var(--ic-action-destructive);
}

ic-button:focus-within .menu-item-label,
ic-button:focus-within .shortcut,
ic-button:focus-within .menu-item-description {
  color: var(--ic-color-white-text);
}

:host(.disabled) ic-button:focus-within .focus-border {
  border-color: var(--ic-architectural-white);
}

:host(.disabled) ic-button:focus-within .menu-item-label,
:host(.disabled) ic-button:focus-within .shortcut,
:host(.disabled) ic-button:focus-within .menu-item-description {
  color: var(--ic-theme-lighten-40) !important;
}

:host::part(button):active {
  background-color: var(--ic-action-dark-bg-active);
  color: var(--ic-color-primary-text) !important;
}

:host([variant="destructive"])::part(button):active {
  background-color: var(--ic-action-dark-bg-active);
  color: var(--ic-action-destructive) !important;
}

.menu-item-info {
  display: flex;
  flex-direction: column;
  flex: 1;
  align-items: baseline;
  white-space: pre-line;
  text-align: left;
}

.menu-labels {
  display: flex;
  align-items: baseline;
  gap: var(--ic-space-xs);
}

.menu-item-description {
  color: var(--ic-color-secondary-text);
}

.shortcut {
  color: var(--ic-color-tertiary-text);
  display: flex;
  flex: min-content;
}

:host([variant="destructive"]) .menu-item-label {
  text-transform: uppercase;
  font-weight: 600;
  letter-spacing: 0.025em;
}

ic-button:active .menu-item-label {
  color: var(--ic-color-primary-text);
}

ic-button:active .menu-item-description {
  color: var(--ic-color-secondary-text);
}

ic-button:active .shortcut {
  color: var(--ic-color-tertiary-text);
}

:host([variant="destructive"]) ic-button:active .menu-item-label {
  color: var(--ic-action-destructive) !important;
}

.icon,
.check-icon,
.submenu-icon,
.icon ::slotted(svg),
.check-icon svg,
.submenu-icon svg {
  width: var(--ic-space-lg);
  height: var(--ic-space-lg);
}

.check-icon.hide {
  visibility: hidden;
}

/** High Contrast **/
@media (forced-colors: active) {
  :host::part(button) {
    border: canvas;
  }

  :host::part(button):focus-visible {
    outline: var(--ic-space-xxxs) solid transparent;
  }

  .focus-border {
    border-color: canvas;
  }

  :host(.disabled) ic-button:focus-within .focus-border {
    border-color: GrayText;
  }

  :host(.disabled) .menu-item-info,
  :host(.disabled) .menu-item-label,
  :host(.disabled) .shortcut,
  :host(.disabled) .menu-item-description,
  :host(.disabled) ic-button:focus-within .menu-item-info,
  :host(.disabled) ic-button:focus-within .menu-item-label,
  :host(.disabled) ic-button:focus-within .shortcut,
  :host(.disabled) ic-button:focus-within .menu-item-description {
    color: GrayText !important;
  }

  :host(.disabled)::part(button):focus-visible,
  :host(.disabled)::part(button):focus {
    outline-color: GrayText !important;
  }
}
