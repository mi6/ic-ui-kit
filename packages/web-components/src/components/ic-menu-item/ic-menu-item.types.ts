export type IcMenuItemVariants = "toggle" | "destructive" | "default";
