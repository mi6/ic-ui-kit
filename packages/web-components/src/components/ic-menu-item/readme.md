# ic-menu-item



<!-- Auto Generated Below -->


## Properties

| Property             | Attribute             | Description                                                                                                                                  | Type                                                                                                                                                                                     | Default     |
| -------------------- | --------------------- | -------------------------------------------------------------------------------------------------------------------------------------------- | ---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- | ----------- |
| `description`        | `description`         | The description displayed in the menu item, below the label.                                                                                 | `string`                                                                                                                                                                                 | `undefined` |
| `disabled`           | `disabled`            | If `true`, the menu item will be in disabled state.                                                                                          | `boolean`                                                                                                                                                                                | `false`     |
| `href`               | `href`                | The URL that the link points to. This will render the menu item as an "a" tag.                                                               | `string`                                                                                                                                                                                 | `undefined` |
| `hreflang`           | `hreflang`            | The human language of the linked URL.                                                                                                        | `string`                                                                                                                                                                                 | `undefined` |
| `keyboardShortcut`   | `keyboard-shortcut`   | The label describing the keyboard shortcut for a menu item's action.                                                                         | `string`                                                                                                                                                                                 | `undefined` |
| `label` _(required)_ | `label`               | The label to display in the menu item.                                                                                                       | `string`                                                                                                                                                                                 | `undefined` |
| `referrerpolicy`     | `referrerpolicy`      | How much of the referrer to send when following the link.                                                                                    | `"" \| "no-referrer" \| "no-referrer-when-downgrade" \| "origin" \| "origin-when-cross-origin" \| "same-origin" \| "strict-origin" \| "strict-origin-when-cross-origin" \| "unsafe-url"` | `undefined` |
| `rel`                | `rel`                 | The relationship of the linked URL as space-separated link types.                                                                            | `string`                                                                                                                                                                                 | `undefined` |
| `submenuTriggerFor`  | `submenu-trigger-for` | This references the popover menu instance that the menu item is a trigger for. If this prop is set, then the variant will always be default. | `string`                                                                                                                                                                                 | `undefined` |
| `target`             | `target`              | The place to display the linked URL, as the name for a browsing context (a tab, window, or iframe).                                          | `string`                                                                                                                                                                                 | `undefined` |
| `variant`            | `variant`             | The variant of the menu item.                                                                                                                | `"default" \| "destructive" \| "toggle"`                                                                                                                                                 | `"default"` |


## Slots

| Slot     | Description                                                |
| -------- | ---------------------------------------------------------- |
| `"icon"` | Content will be placed to the left of the menu item label. |


## Dependencies

### Used by

 - [ic-popover-menu](../ic-popover-menu)

### Depends on

- [ic-typography](../ic-typography)
- [ic-button](../ic-button)

### Graph
```mermaid
graph TD;
  ic-menu-item --> ic-typography
  ic-menu-item --> ic-button
  ic-button --> ic-loading-indicator
  ic-button --> ic-tooltip
  ic-loading-indicator --> ic-typography
  ic-tooltip --> ic-typography
  ic-popover-menu --> ic-menu-item
  style ic-menu-item fill:#f9f,stroke:#333,stroke-width:4px
```

----------------------------------------------


