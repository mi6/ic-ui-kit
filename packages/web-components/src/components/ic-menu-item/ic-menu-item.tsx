import {
  Component,
  Element,
  Host,
  Prop,
  h,
  State,
  Event,
  EventEmitter,
  Listen,
} from "@stencil/core";
import {
  getParentElement,
  isSlotUsed,
  onComponentRequiredPropUndefined,
  removeDisabledFalse,
} from "../../utils/helpers";
import { IcMenuItemVariants } from "./ic-menu-item.types";
import Check from "../../assets/check-icon.svg";
import Chevron from "../../assets/chevron-icon.svg";

/**
 * @slot icon - Content will be placed to the left of the menu item label.
 */

@Component({
  tag: "ic-menu-item",
  styleUrl: "ic-menu-item.css",
  shadow: {
    delegatesFocus: true,
  },
})
export class MenuItem {
  @Element() el: HTMLIcMenuItemElement;

  @State() toggleChecked: boolean = false;

  /**
   * The description displayed in the menu item, below the label.
   */
  @Prop() description?: string;

  /**
   * If `true`, the menu item will be in disabled state.
   */
  @Prop() disabled?: boolean = false;

  /**
   * The URL that the link points to. This will render the menu item as an "a" tag.
   */
  @Prop() href?: string;

  /**
   * The human language of the linked URL.
   */
  @Prop() hreflang?: string;

  /**
   * The label describing the keyboard shortcut for a menu item's action.
   */
  @Prop() keyboardShortcut?: string;

  /**
   * The label to display in the menu item.
   */
  @Prop() label!: string;

  /**
   * How much of the referrer to send when following the link.
   */
  @Prop() referrerpolicy?: ReferrerPolicy;

  /**
   * The relationship of the linked URL as space-separated link types.
   */
  @Prop() rel?: string;

  /**
   * This references the popover menu instance that the menu item is a trigger for. If this prop is set, then the variant will always be default.
   */
  @Prop() submenuTriggerFor?: string;

  /**
   * The place to display the linked URL, as the name for a browsing context (a tab, window, or iframe).
   */
  @Prop() target?: string;

  /**
   * The variant of the menu item.
   */
  @Prop({ mutable: true, reflect: true }) variant: IcMenuItemVariants =
    "default";

  /**
   * @internal Emitted when item loses focus.
   */
  @Event() childBlur: EventEmitter<void>;

  /**
   * @internal Emitted when the user clicks a menu item.
   */
  @Event() handleMenuItemClick: EventEmitter<{
    label: string;
    hasSubMenu: boolean;
  }>;

  /**
   * @internal Emitted when the user clicks a menu item that triggers a popover menu instance.
   */
  @Event() triggerPopoverMenuInstance: EventEmitter<void>;

  componentWillLoad(): void {
    // This ensures that trigger menu items are always set to the default variant
    if (this.submenuTriggerFor !== undefined && this.variant !== "default") {
      this.variant = "default";
    }
    removeDisabledFalse(this.disabled, this.el);
  }

  componentDidLoad(): void {
    onComponentRequiredPropUndefined(
      [{ prop: this.label, propName: "label" }],
      "Menu Item"
    );
  }

  @Listen("click", { capture: true })
  handleHostClick(e: Event): void {
    if (this.disabled) {
      e.stopImmediatePropagation();
    }
  }

  private handleClick = (e: Event): void => {
    e.preventDefault();
    if (this.submenuTriggerFor !== undefined) {
      this.triggerPopoverMenuInstance.emit();
    } else if (this.variant === "toggle") {
      this.toggleChecked
        ? (this.toggleChecked = false)
        : (this.toggleChecked = true);
    }
    this.handleMenuItemClick.emit({
      label: this.label,
      hasSubMenu: !!this.el.submenuTriggerFor,
    });
  };

  private getMenuItemAriaLabel = (): string => {
    let ariaLabel = this.label;

    if (this.description !== undefined) {
      ariaLabel = `${ariaLabel}, ${this.description}`;
    }

    if (this.keyboardShortcut !== undefined) {
      ariaLabel = `${ariaLabel}, ${this.keyboardShortcut}`;
    }

    if (this.variant === "destructive") {
      ariaLabel = `${ariaLabel}, destructive`;
    }

    if (this.submenuTriggerFor !== undefined) {
      ariaLabel = `${ariaLabel}, triggers submenu`;
    }

    if (this.el.classList.contains("ic-popover-submenu-back-button")) {
      ariaLabel = "Go back to parent menu";
    }
    const parentEl = getParentElement(this.el);

    if (
      parentEl.tagName === "IC-MENU-GROUP" &&
      (parentEl as HTMLIcMenuGroupElement).label
    ) {
      return `${ariaLabel}, ${
        (parentEl as HTMLIcMenuGroupElement).label
      } menu group`;
    } else {
      return ariaLabel;
    }
  };

  render() {
    // A helper function that checks if a prop has been defined
    const isPropDefined = (prop: string) => {
      return prop !== undefined ? prop : null;
    };

    // A sub-component to layout the menu information correctly in ic-button
    const MenuItemInformation = () => {
      return (
        <div class="menu-item-info">
          <div class="menu-labels">
            <ic-typography class="menu-item-label">{this.label}</ic-typography>
            {this.keyboardShortcut && (
              <ic-typography variant="caption" class="shortcut">
                {this.keyboardShortcut}
              </ic-typography>
            )}
          </div>
          {this.description && (
            <ic-typography class="menu-item-description" variant="caption">
              {this.description}
            </ic-typography>
          )}
        </div>
      );
    };

    return (
      <Host
        class={{
          ["disabled"]: this.disabled,
        }}
      >
        <li
          role={this.variant === "toggle" ? "menuitemcheckbox" : "menuitem"}
          aria-disabled={`${this.disabled}`}
          aria-checked={
            this.variant === "toggle" && this.toggleChecked === true
              ? true
              : this.variant === "toggle" && this.toggleChecked === false
              ? false
              : undefined
          }
        >
          <ic-button
            disableTooltip
            fullWidth
            variant="tertiary"
            onClick={this.handleClick}
            href={isPropDefined(this.href)}
            hreflang={isPropDefined(this.hreflang)}
            target={isPropDefined(this.target)}
            rel={isPropDefined(this.rel)}
            referrerpolicy={
              this.referrerpolicy !== undefined ? this.referrerpolicy : null
            }
            aria-disabled={`${this.disabled}`}
            aria-label={this.getMenuItemAriaLabel()}
            aria-controls={
              this.submenuTriggerFor !== undefined
                ? `ic-popover-submenu-${this.submenuTriggerFor}`
                : false
            }
            aria-haspopup={
              this.submenuTriggerFor !== undefined ||
              this.el.classList.contains("ic-popover-submenu-back-button")
                ? "menu"
                : false
            }
            aria-owns={
              this.submenuTriggerFor !== undefined
                ? `ic-popover-submenu-${this.submenuTriggerFor}`
                : false
            }
          >
            <div class="focus-border">
              {isSlotUsed(this.el, "icon") && (
                <span class="icon">
                  <slot name="icon"></slot>
                </span>
              )}
              <MenuItemInformation />
              {this.variant === "toggle" && (
                <span
                  class={{
                    ["check-icon"]: true,
                    ["hide"]: !this.toggleChecked,
                  }}
                  aria-hidden="true"
                  innerHTML={Check}
                />
              )}
              {this.submenuTriggerFor !== undefined && (
                <span
                  class={{ ["submenu-icon"]: true }}
                  aria-hidden="true"
                  innerHTML={Chevron}
                />
              )}
            </div>
          </ic-button>
        </li>
      </Host>
    );
  }
}
