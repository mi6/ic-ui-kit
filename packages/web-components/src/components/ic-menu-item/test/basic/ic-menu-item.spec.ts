import { newSpecPage } from "@stencil/core/testing";
import { MenuItem } from "../../ic-menu-item";
import { Button } from "../../../ic-button/ic-button";

describe("menu item variants", () => {
  it("should render the default variant", async () => {
    const page = await newSpecPage({
      components: [MenuItem],
      html: `<ic-menu-item
            label="Default variant"
          />`,
    });

    expect(page.root).toMatchSnapshot();
  });

  it("should render a menu item with a description", async () => {
    const page = await newSpecPage({
      components: [MenuItem],
      html: `<ic-menu-item
            label="Default variant"
            description="This is the default variant of the menu item with a description"
          />`,
    });

    expect(page.root).toMatchSnapshot();
  });

  it("should render with keyboard shortcut", async () => {
    const page = await newSpecPage({
      components: [MenuItem],
      html: `<ic-menu-item
              label="Toggle variant"
              keyboard-shortcut="Cmd+"
            />`,
    });
    expect(page.root).toMatchSnapshot();
  });

  it("should render the toggle variant", async () => {
    const page = await newSpecPage({
      components: [MenuItem, Button],
      html: `<ic-menu-item
            variant="toggle"
            label="Toggle variant"
            id="test-menu-item"
          />`,
    });

    expect(page.root).toMatchSnapshot();
    expect(page.rootInstance.variant).toMatch("toggle");
    expect(page.rootInstance.toggleChecked).toBeFalsy();

    const button = page.root.shadowRoot
      .querySelector("li > ic-button")
      .shadowRoot.querySelector("button");

    button.click();
    await page.waitForChanges;
  });

  it("should render the destructive variant", async () => {
    const page = await newSpecPage({
      components: [MenuItem],
      html: `<ic-menu-item
            variant="destuctive"
            label="Destructive variant"
          />`,
    });

    expect(page.root).toMatchSnapshot();
  });

  it("should render a menu item that triggers a popover menu instance", async () => {
    const page = await newSpecPage({
      components: [MenuItem],
      html: `<ic-menu-item
            variant="destuctive"
            submenu-trigger-for="submenu-1"
            label="Destructive variant"
          />`,
    });

    expect(page.rootInstance.submenuTriggerFor).not.toBeUndefined();

    expect(page.rootInstance.variant).toMatch("default");
  });

  it('should emit the triggerPopoverMenuInstance event when the menu item has the prop: "submenu-trigger-for" and is clicked', async () => {
    const page = await newSpecPage({
      components: [MenuItem, Button],
      html: `<ic-menu-item
      id="test-menu-item"
            submenu-trigger-for="submenu-1"
            label="I emit an event"
          />`,
    });

    const eventSpy = jest.fn();
    page.win.addEventListener("triggerPopoverMenuInstance", eventSpy);
    expect(page.rootInstance.submenuTriggerFor).not.toBeUndefined();

    const element = await document.getElementById("test-menu-item");

    await element.click();

    await page.waitForChanges();

    await page.rootInstance.handleClick;
  });

  it("should prevent default action on click", async () => {
    const page = await newSpecPage({
      components: [MenuItem],
      html: `<ic-menu-item
      id="test-menu-item"
            label="I emit an event"
            submenu-trigger-for="test-submenu"
          />`,
    });

    const mockEvent = { preventDefault: jest.fn() };
    page.rootInstance.handleClick(mockEvent as unknown as MouseEvent);

    expect(mockEvent.preventDefault).toHaveBeenCalled();
  });
});
