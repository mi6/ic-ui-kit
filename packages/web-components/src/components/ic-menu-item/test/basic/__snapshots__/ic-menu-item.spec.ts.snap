// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`menu item variants should render a menu item with a description 1`] = `
<ic-menu-item description="This is the default variant of the menu item with a description" label="Default variant" variant="default">
  <mock:shadow-root>
    <li aria-disabled="false" role="menuitem">
      <ic-button aria-disabled="false" aria-label="Default variant, This is the default variant of the menu item with a description" disabletooltip="" fullwidth="" variant="tertiary">
        <div class="focus-border">
          <div class="menu-item-info">
            <div class="menu-labels">
              <ic-typography class="menu-item-label">
                Default variant
              </ic-typography>
            </div>
            <ic-typography class="menu-item-description" variant="caption">
              This is the default variant of the menu item with a description
            </ic-typography>
          </div>
        </div>
      </ic-button>
    </li>
  </mock:shadow-root>
</ic-menu-item>
`;

exports[`menu item variants should render the default variant 1`] = `
<ic-menu-item label="Default variant" variant="default">
  <mock:shadow-root>
    <li aria-disabled="false" role="menuitem">
      <ic-button aria-disabled="false" aria-label="Default variant" disabletooltip="" fullwidth="" variant="tertiary">
        <div class="focus-border">
          <div class="menu-item-info">
            <div class="menu-labels">
              <ic-typography class="menu-item-label">
                Default variant
              </ic-typography>
            </div>
          </div>
        </div>
      </ic-button>
    </li>
  </mock:shadow-root>
</ic-menu-item>
`;

exports[`menu item variants should render the destructive variant 1`] = `
<ic-menu-item label="Destructive variant" variant="destuctive">
  <mock:shadow-root>
    <li aria-disabled="false" role="menuitem">
      <ic-button aria-disabled="false" aria-label="Destructive variant" disabletooltip="" fullwidth="" variant="tertiary">
        <div class="focus-border">
          <div class="menu-item-info">
            <div class="menu-labels">
              <ic-typography class="menu-item-label">
                Destructive variant
              </ic-typography>
            </div>
          </div>
        </div>
      </ic-button>
    </li>
  </mock:shadow-root>
</ic-menu-item>
`;

exports[`menu item variants should render the toggle variant 1`] = `
<ic-menu-item id="test-menu-item" label="Toggle variant" variant="toggle">
  <mock:shadow-root>
    <li aria-disabled="false" role="menuitemcheckbox">
      <ic-button class="button-size-default button-variant-tertiary full-width" exportparts="button">
        <mock:shadow-root>
          <button aria-disabled="false" aria-label="Toggle variant" class="button" part="button" type="button">
            <slot></slot>
          </button>
        </mock:shadow-root>
        <div class="focus-border">
          <div class="menu-item-info">
            <div class="menu-labels">
              <ic-typography class="menu-item-label">
                Toggle variant
              </ic-typography>
            </div>
          </div>
          <span aria-hidden="true" class="check-icon hide">
            svg
          </span>
        </div>
      </ic-button>
    </li>
  </mock:shadow-root>
</ic-menu-item>
`;

exports[`menu item variants should render with keyboard shortcut 1`] = `
<ic-menu-item keyboard-shortcut="Cmd+" label="Toggle variant" variant="default">
  <mock:shadow-root>
    <li aria-disabled="false" role="menuitem">
      <ic-button aria-disabled="false" aria-label="Toggle variant, Cmd+" disabletooltip="" fullwidth="" variant="tertiary">
        <div class="focus-border">
          <div class="menu-item-info">
            <div class="menu-labels">
              <ic-typography class="menu-item-label">
                Toggle variant
              </ic-typography>
              <ic-typography class="shortcut" variant="caption">
                Cmd+
              </ic-typography>
            </div>
          </div>
        </div>
      </ic-button>
    </li>
  </mock:shadow-root>
</ic-menu-item>
`;
